File: _sandbox/tests/benchmark_rag_speed.py

import time
import random
import string
import shutil
import logging
from pathlib import Path
from zena_mode.rag_pipeline import LocalRAG

# --- Configuration ---
NUM_DOCS = 100
DOC_LENGTH = 2000 # chars
TEST_DIR = Path("benchmark_rag_data")
DB_PATH = TEST_DIR / "rag_db"

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("RAG_Benchmark")

def generate_synthetic_data(num_docs, length):
    """Generate random documents to simulate load."""
    docs = []
    logger.info(f"Generating {num_docs} synthetic documents...")
    for i in range(num_docs):
        # Generate random text
        text = ''.join(random.choices(string.ascii_letters + " " * 10, k=length))
        # Insert some searchable keywords
        if i % 10 == 0:
            text += " The secret password is BlueSky."
        docs.append({
            "content": text,
            "url": f"http://fake.com/doc_{i}",
            "title": f"Synthetic Doc {i}"
        })
    return docs

def run_benchmark():
    # Setup
    if TEST_DIR.exists():
        shutil.rmtree(TEST_DIR)
    TEST_DIR.mkdir()

    # Data
    docs = generate_synthetic_data(NUM_DOCS, DOC_LENGTH)

    # Init RAG
    rag = LocalRAG(cache_dir=DB_PATH)
    
    # Measure Ingest
    logger.info("--- Starting Ingest Benchmark ---")
    start_time = time.time()
    rag.build_index(docs)
    end_time = time.time()
    
    ingest_duration = end_time - start_time
    docs_per_sec = NUM_DOCS / ingest_duration
    
    # Check Storage Size
    rag.save(DB_PATH)
    total_size = sum(f.stat().st_size for f in DB_PATH.glob('**/*') if f.is_file()) / (1024*1024)
    
    logger.info(f"Ingest Complete: {ingest_duration:.2f}s")
    logger.info(f"Throughput: {docs_per_sec:.2f} docs/sec")
    logger.info(f"Storage Size: {total_size:.2f} MB")

    # Measure Search
    logger.info("--- Starting Search Benchmark ---")
    search_start = time.time()
    results = rag.search("secret password BlueSky", k=3)
    search_end = time.time()
    
    search_latency = (search_end - search_start) * 1000 # ms
    logger.info(f"Search Latency: {search_latency:.2f} ms")
    logger.info(f"Found: {len(results)} results")
    if results:
         logger.info(f"Top Result Score: {results[0].get('score', 0):.4f}")

    # Report
    print(f"\n=== BENCHMARK RESULTS (Baseline) ===")
    print(f"Documents: {NUM_DOCS}")
    print(f"Total Time: {ingest_duration:.4f} s")
    print(f"Throughput: {docs_per_sec:.2f} docs/s")
    print(f"Search Latency: {search_latency:.2f} ms")
    print(f"Storage: {total_size:.2f} MB")
    print("====================================")

    # Cleanup
    # shutil.rmtree(TEST_DIR)

if __name__ == "__main__":
    run_benchmark()

File: rust_output/src/benchmark_rag_speed.rs
use anyhow::{Result, Context};
use crate::rag_pipeline::{LocalRAG};
use std::collections::HashMap;

pub const NUM_DOCS: i64 = 100;

pub const DOC_LENGTH: i64 = 2000;

pub static TEST_DIR: std::sync::LazyLock<PathBuf> = std::sync::LazyLock::new(|| Default::default());

pub const DB_PATH: &str = "TEST_DIR / 'rag_db";

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Generate random documents to simulate load.
pub fn generate_synthetic_data(num_docs: String, length: String) -> () {
    // Generate random documents to simulate load.
    let mut docs = vec![];
    logger.info(format!("Generating {} synthetic documents...", num_docs));
    for i in 0..num_docs.iter() {
        let mut text = random.choices((string.ascii_letters + (" ".to_string() * 10)), /* k= */ length).join(&"".to_string());
        if (i % 10) == 0 {
            text += " The secret password is BlueSky.".to_string();
        }
        docs.push(HashMap::from([("content".to_string(), text), ("url".to_string(), format!("http://fake.com/doc_{}", i)), ("title".to_string(), format!("Synthetic Doc {}", i))]));
    }
    docs
}

/// Run benchmark.
pub fn run_benchmark() -> () {
    // Run benchmark.
    if TEST_DIR.exists() {
        std::fs::remove_dir_all(TEST_DIR).ok();
    }
    TEST_DIR.create_dir_all();
    let mut docs = generate_synthetic_data(NUM_DOCS, DOC_LENGTH);
    let mut rag = LocalRAG(/* cache_dir= */ DB_PATH);
    logger.info("--- Starting Ingest Benchmark ---".to_string());
    let mut start_time = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    rag.build_index(docs);
    let mut end_time = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    let mut ingest_duration = (end_time - start_time);
    let mut docs_per_sec = (NUM_DOCS / ingest_duration);
    rag.save(DB_PATH);
    let mut total_size = (DB_PATH.glob("**/*".to_string()).iter().filter(|f| f.is_file()).map(|f| f.stat().st_size).collect::<Vec<_>>().iter().sum::<i64>() / (1024 * 1024));
    logger.info(format!("Ingest Complete: {:.2}s", ingest_duration));
    logger.info(format!("Throughput: {:.2} docs/sec", docs_per_sec));
    logger.info(format!("Storage Size: {:.2} MB", total_size));
    logger.info("--- Starting Search Benchmark ---".to_string());
    let mut search_start = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    let mut results = rag.search("secret password BlueSky".to_string(), /* k= */ 3);
    let mut search_end = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
    let mut search_latency = ((search_end - search_start) * 1000);
    logger.info(format!("Search Latency: {:.2} ms", search_latency));
    logger.info(format!("Found: {} results", results.len()));
    if results {
        logger.info(format!("Top Result Score: {:.4}", results[0].get(&"score".to_string()).cloned().unwrap_or(0)));
    }
    println!("\n=== BENCHMARK RESULTS (Baseline) ===");
    println!("Documents: {}", NUM_DOCS);
    println!("Total Time: {:.4} s", ingest_duration);
    println!("Throughput: {:.2} docs/s", docs_per_sec);
    println!("Search Latency: {:.2} ms", search_latency);
    println!("Storage: {:.2} MB", total_size);
    println!("{}", "====================================".to_string());
}
